### PR TITLE
Add aria-label for notebook cells to improve screen-reader visibility

### PR DESCRIPTION
**Pull request by KrishnaKumarHariprasannan**
## References
https://github.com/jupyterlab/jupyterlab/issues/10347

## Code changes

Changes related to adding aria-label for different types of notebook cells and related references to access translator

## User-facing changes

Changes only visible to screenreaders

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by KrishnaKumarHariprasannan** [comment]
@manfromjupyter, thanks a lot for waiting on this. Can you please check if these changes look good on a screen-reader? I'll open it up for review once you give me the go-ahead (and after I update the tests).

File: packages/cells/src/widget.ts [hunk 1]
```diff
@@ -1420,6 +1468,10 @@ export class MarkdownCell extends AttachmentsCell<IMarkdownCellModel> {
   constructor(options: MarkdownCell.IOptions) {
     super(options);
     this.addClass(MARKDOWN_CELL_CLASS);
+    this.node.setAttribute(
+      'aria-label',
+      this.transBundle.__(MARKDOWN_CELL_ARIA_LABEL)
```
Comment (krassowski) [inline packages/cells/src/widget.ts:1473]: Quick comment: for the translation extractor to work it needs to match a specific regular expression and it includes `trans.__('string')` and the actual string needs to be inside of the the parentheses. I know it may sound disappointing that the translation tooling is not advanced enough to figure out the strings during runtime (well that would be difficult anyways) or using static analysis (this could work but apparently it was not written yet or did not become mainstream).

Long story short you will need to rename `transBundle` to `trans` and use the strings directly rather than referring to them by constants (or define the constants somewhere where you have access to `trans`).

Comment (KrishnaKumarHariprasannan) [reply packages/cells/src/widget.ts:1473]: Thanks for the comment @krassowski, and for saving me a few hours of debugging! Is there any place this is documented (I checked READMEs in `@jupyterlab/translation*` and [API docs](https://jupyterlab.readthedocs.io/en/stable/api/modules/translation.html))? If not, can you please point me to where this can be added?

Comment (krassowski) [reply packages/cells/src/widget.ts:1473]: I'm afraid the docs might be missing at this stage. @fcollonval and @goanpeca any thoughts?

Comment (fcollonval) [reply packages/cells/src/widget.ts:1473]: We definitely need to add that documentation. The best place seems the developer documentation of JupyterLab.

File: packages/cells/src/widget.ts [hunk 2]
```diff
@@ -1032,6 +1055,11 @@ export class CodeCell extends Cell<ICodeCellModel> {
   private _outputLengthHandler(sender: OutputArea, args: number) {
     const force = args === 0 ? true : false;
     this.toggleClass(NO_OUTPUTS_CLASS, force);
+    const trans = this.translator.load('jupyterlab');
+    const ariaLabel = force
+      ? 'Code Cell Content'
+      : 'Code Cell Content with Output';
+    this.node.setAttribute('aria-label', trans.__(ariaLabel));
```
Comment (fcollonval) [inline packages/cells/src/widget.ts:1062]:
```suggestion
    const ariaLabel = force
      ? trans.__('Code Cell Content')
      : trans.__('Code Cell Content with Output');
    this.node.setAttribute('aria-label', ariaLabel);
```

File: packages/cells/src/widget.ts [hunk 3]
```diff
@@ -716,12 +729,19 @@ export class CodeCell extends Cell<ICodeCellModel> {
...
     // Only save options not handled by parent constructor.
     const rendermime = (this._rendermime = options.rendermime);
     const contentFactory = this.contentFactory;
     const model = this.model;
 
+    // Note that modifying the below label warrants one to also modify
+    // the same in this._outputLengthHandler. Ideally, this label must
+    // have been a constant and used in both places but it is not done
+    // so because of limitations in the translation manager.
+    let ariaLabel = 'Code Cell Content with Output';
```
Comment (fcollonval) [inline packages/cells/src/widget.ts:743]: it feel unnatural as mentioned by @krassowski. But you need to wrap the string with the translation method for it to be extracted from the code. It does not find the strings if you request the translation of a variable.

```suggestion
    let ariaLabel = trans.__('Code Cell Content with Output');
```

File: packages/cells/src/widget.ts [hunk 4]
```diff
@@ -740,6 +760,7 @@ export class CodeCell extends Cell<ICodeCellModel> {
       // if there are no outputs.
       if (model.outputs.length === 0) {
         this.addClass(NO_OUTPUTS_CLASS);
+        ariaLabel = 'Code Cell Content';
```
Comment (fcollonval) [inline packages/cells/src/widget.ts:763]:
```suggestion
        ariaLabel = trans.__('Code Cell Content');
```

File: packages/cells/src/widget.ts [hunk 5]
```diff
@@ -755,6 +776,7 @@ export class CodeCell extends Cell<ICodeCellModel> {
       });
     }
     model.stateChanged.connect(this.onStateChanged, this);
+    this.node.setAttribute('aria-label', trans.__(ariaLabel));
```
Comment (fcollonval) [inline packages/cells/src/widget.ts:779]:
```suggestion
    this.node.setAttribute('aria-label', ariaLabel);
```

File: packages/cells/src/widget.ts (hunk 1 above)
Comment (KrishnaKumarHariprasannan) [reply packages/cells/src/widget.ts]: @krassowski , I've resolved this comment - please check and add any other feedback you have. In my cursory search, I wasn't able to find any tests targeting aria content - let me know how we should proceed in this regard.

> Quick comment: for the translation extractor to work it needs to match a specific regular expression and it includes trans.__('string')

@krassowski / @fcollonval , can you please point me to this regex/related module ? I'll have a first-pass at adding documentation with this information.

**Review by fcollonval — CHANGES_REQUESTED** [review]
Hey @KrishnaKumarHariprasannan I added some changes to handle properly internationalization.

File: packages/cells/src/widget.ts (hunk 3 above)
Comment (KrishnaKumarHariprasannan) [reply packages/cells/src/widget.ts]: Thanks for the super-fast review @fcollonval !

Oops. From the previous comment, `define the constants somewhere where you have access to trans`, I assumed this would work if the variables are reachable by `trans`. I've added all your suggestions with the recent commit.

File: tsconfigdoc.json [hunk 6]
```diff
@@ -234,6 +234,9 @@
     {
       "path": "./packages/statusbar-extension"
     },
+    {
+      "path": "./packages/tabmanager-extension"
+    },
```
Comment (fcollonval) [inline tsconfigdoc.json:239]: This is the root cause for the CI errors

```suggestion
```

**Review by fcollonval — APPROVED** [review]
LGTM thanks @KrishnaKumarHariprasannan

**Comment by fcollonval** [comment]
@krassowski @blink1073  should this be backported to 3.1.x?

**Comment by blink1073** [comment]
> should this be backported to 3.1.x?

I'd say no given that there is an (optional) API change.

**Comment by manfromjupyter** [comment]
@KrishnaKumarHariprasannan super sorry for the delay on testing this. I did finally check it out but I assume since it's been merged that the binder no longer works. Getting the following error. Anyway will test it once it comes out I suppose. Thank you again for working on this.
testing it binder issues

**Comment by fcollonval** [comment]
@manfromjupyter you could try on the master branch with: https://mybinder.org/v2/gh/jupyterlab/jupyterlab/master?urlpath=lab-dev

This is what I get from Firefox Accessibility tool: 
image

**Comment by manfromjupyter** [comment]
@fcollonval perfect! Thank you.

@KrishnaKumarHariprasannan so TLDR works great! For more detail here's what I see.
1. When entering through a a markdown cell, I hear "Markdown content cell edit type of text" which is perfect. Tells me what type of cell, that I can edit it currently, and what format the editing accepts. Then if I down arrow to read next element in the order it reads me the entire cell which both of things together is a perfect for the user! Thank you
2. When entering through a a raw cell, I hear "Raw cell edit type of text" which is perfect. Tells me what type of cell, that I can edit it currently, and what format the editing accepts. Then if I down arrow to read next element in the order it reads me the entire cell which once again is a perfect experience for the user! Thank you
3. For code cells it wants to read the "[ ]:" characters first which is annoying but I understand it's purpose. Should have an aria-label instead saying what that means. Aka "Unexecuted Code Cell" and then once a number gets in there (ex. "3"), it changes to "Code Cell with execution order "3". Then after next element in the screenreader order, then your aria-label is applied despite it appearing in the html around both of these elements. It says "Code Cell Content edit type of text" which is the same format above and works perfectly.

When the notebook cell has output though, I only see 1-2 tiny issues. When going in forward (normal) order over the very first cell IF it is a code cell that has already been run (so has output), the notebook section only cays "edit". Then when I get to the output it just reads the output without telling me it was the output of the code cell prior (or "Output of code cell with execution order 3:"). However when I go to the following code cell block and then return backwards to the one before, it says "Code Cell Content with Output edit type of text". So a little bizarre, definitely not what I'd expect by looking at the HTML given the position of your aria-labels. When I view other code cells with output it works fine though... The output of that cell and others still just reads the text without any additional information, but that may be by design. That might be fine but might be misunderstood as markdown text, or more of the same text/code from the code cell... I don't know. If it was me, I might want a little label thrown before the text that spits out that says my above example of "Output of code cell with execution order 3:". As per the first code cell problem I do not know why it is causing that, but I suspect the "notebook actions notebook region end" is kind of overriding it because it is right under it. Given the request of the ticket, I feel you'd done a job well done and works; these just might be things to build onto it in the future. 

## Non-needed things for WCAG2.1 but additional extra credit things that could help with their user experience in the future
1. In all code cell types, say cell block number that is, but I recognize it's super high lift and it might get confusing when also adding the run order number to the code cells also. Could say "Cell 1, type Code Cell, with execution order '3'"?
2. Add aria-label to the "[ ]:", "[3]:", etc element to instead something like "Unexecuted Code Cell", "Code Cell with execution order "3", etc.
3. Add label prefix to output, ex: "Output of code cell with execution order 3:"
4. First cell doesn't read what type it is if it is a code cell with output (when reading in normal order). May just need a random html element like a screenreader only hr so: `<hr class="sr-only"></hr>` there to separate the outer element with the region and the div of the first cell. Some customers may start their notebook with a heading (H1) or something else so will be fine for them, just code cells will not say what kind for these screenreader users.

Excuse any type-os. Only did a single pass.

**Comment by KrishnaKumarHariprasannan** [comment]
Thanks for checking this @manfromjupyter ! I considered adding more context to the label for code cells and I figured cell IDs would be more appropriate than using execution order (execution orders are unordered and not unique - single cell can assume multiple execution order in its lifetime). Cell IDs are additional fields added to each cell with notebooks using nbformat>=4.5 and I came across https://github.com/jupyterlab/jupyterlab/pull/10018#issuecomment-809587175 which outlines the plan to upgrade jlab to use nbformat>=4.5 by default for all new notebooks created. This would mean we can tie together cells with their corresponding outputs using cell IDs (with the caveat being the IDs being content hashes and so likely not human-friendly). I propose we open an issue and implement it after the said upgrade. Happy to hear others' thoughts on this.

Re. the issue with the first code cell with output, I'll check for anything obvious that might contribute to this and revert here with my findings.
